utils/rate_limiter.go

```diff
@@ -62,8 +65,9 @@ func (b *rateLimiter) CheckCredit(itemCost float64) bool {
 	b.lastTick = currentTime
 	// calculate how much credit have we accumulated since the last tick
 	b.balance += elapsedTime.Seconds() * b.creditsPerSecond
-	if b.balance > b.creditsPerSecond {
-		b.balance = b.creditsPerSecond
+	upperbound := math.Max(b.creditsPerSecond, itemCost)
```

Comment:
somehow this feels dirty. I.e. it is not obvious why this value should be the upper bound in the general case. What if we made this into a `maxBalance` parameter instead?